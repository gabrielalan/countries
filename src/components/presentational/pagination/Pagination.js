import React from 'react';
import PropTypes from 'prop-types';
import { classNames } from '../../../functions/elementHelpers';

function Pagination({ current, limit, total, onChange }) {
  const buttons = [];
  const numberOfPages = Math.ceil(total / limit);
  const isFirstPage = current <= 0;
  const isLastPage = (numberOfPages - 1) <= current;

  for (let i = 0; i < numberOfPages; i++) {
    buttons.push(<li key={i} className={classNames({ 'pagination__page': true, 'pagination__page--current': current === i })}>
      <button type="button" onClick={() => current !== i && onChange(i)} disabled={current === i}>{i + 1}</button>
    </li>);
  }

  return <nav className="pagination">
    <ul className="pagination__wrapper">
      <li className={classNames({ 'pagination__page': true, 'pagination__page--disabled': isFirstPage } )}>
        <button type="button" onClick={() => !isFirstPage && onChange(current - 1)} disabled={isFirstPage}>previous</button>
      </li>
      {buttons}
      <li className={classNames({ 'pagination__page': true, 'pagination__page--disabled': isLastPage } )}>
        <button type="button" onClick={() => !isLastPage && onChange(current + 1)} disabled={isLastPage}>next</button>
      </li>
    </ul>
  </nav>;
}

Pagination.defaultProps = {
  onChange: () => null
};

Pagination.propTypes = {
  current: PropTypes.number.isRequired,
  limit: PropTypes.number.isRequired,
  total: PropTypes.number.isRequired,
  onChange: PropTypes.func,
};

export default React.memo(Pagination);