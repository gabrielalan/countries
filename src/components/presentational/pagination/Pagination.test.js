import React from 'react';
import renderer from 'react-test-renderer';
import Pagination from './Pagination';

describe('Pagination', () => {
  it('should render correct number of pages', () => {
    const tenPages = renderer.create(<Pagination current={0} limit={10} total={100} />).toJSON();
    expect(tenPages).toMatchSnapshot();

    const elevenPages = renderer.create(<Pagination current={0} limit={10} total={101} />).toJSON();
    expect(elevenPages).toMatchSnapshot();
  });

  it('should render correct number of pages', () => {
    const spy = jest.fn();
    const { root } = renderer.create(<Pagination current={0} limit={1} total={5} onChange={spy} />);

    const buttons = root.findAllByType('button');

    buttons[0].props.onClick();
    buttons[1].props.onClick();
    expect(spy).not.toHaveBeenCalled();
    
    buttons[2].props.onClick();
    expect(spy).toHaveBeenCalled();
  });
});