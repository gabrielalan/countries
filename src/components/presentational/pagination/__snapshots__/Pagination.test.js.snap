// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should render correct number of pages 1`] = `
<nav
  className="pagination"
>
  <ul
    className="pagination__wrapper"
  >
    <li
      className="pagination__page pagination__page--disabled"
    >
      <button
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        previous
      </button>
    </li>
    <li
      className="pagination__page pagination__page--current"
    >
      <button
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        1
      </button>
    </li>
    <li
      className="pagination__page"
    >
      <button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        2
      </button>
    </li>
    <li
      className="pagination__page"
    >
      <button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        3
      </button>
    </li>
    <li
      className="pagination__page"
    >
      <button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        4
      </button>
    </li>
    <li
      className="pagination__page"
    >
      <button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        5
      </button>
    </li>
    <li
      className="pagination__page"
    >
      <button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        6
      </button>
    </li>
    <li
      className="pagination__page"
    >
      <button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        7
      </button>
    </li>
    <li
      className="pagination__page"
    >
      <button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        8
      </button>
    </li>
    <li
      className="pagination__page"
    >
      <button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        9
      </button>
    </li>
    <li
      className="pagination__page"
    >
      <button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        10
      </button>
    </li>
    <li
      className="pagination__page"
    >
      <button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        next
      </button>
    </li>
  </ul>
</nav>
`;

exports[`Pagination should render correct number of pages 2`] = `
<nav
  className="pagination"
>
  <ul
    className="pagination__wrapper"
  >
    <li
      className="pagination__page pagination__page--disabled"
    >
      <button
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        previous
      </button>
    </li>
    <li
      className="pagination__page pagination__page--current"
    >
      <button
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        1
      </button>
    </li>
    <li
      className="pagination__page"
    >
      <button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        2
      </button>
    </li>
    <li
      className="pagination__page"
    >
      <button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        3
      </button>
    </li>
    <li
      className="pagination__page"
    >
      <button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        4
      </button>
    </li>
    <li
      className="pagination__page"
    >
      <button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        5
      </button>
    </li>
    <li
      className="pagination__page"
    >
      <button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        6
      </button>
    </li>
    <li
      className="pagination__page"
    >
      <button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        7
      </button>
    </li>
    <li
      className="pagination__page"
    >
      <button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        8
      </button>
    </li>
    <li
      className="pagination__page"
    >
      <button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        9
      </button>
    </li>
    <li
      className="pagination__page"
    >
      <button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        10
      </button>
    </li>
    <li
      className="pagination__page"
    >
      <button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        11
      </button>
    </li>
    <li
      className="pagination__page"
    >
      <button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        next
      </button>
    </li>
  </ul>
</nav>
`;
